File: swift/Sources/ChatGPTOAuth/JWT.swift
import Foundation

/// Contains unverified JWT metadata suitable only for display and request routing, never authorization.
public struct UnverifiedClaims: Sendable, Equatable {
    /// The unverified ChatGPT account routing identifier.
    public let accountID: String?
    /// The unverified plan label for display.
    public let planType: String?
    /// The unverified email address for display.
    public let email: String?

    /// Creates a set of unverified display and routing claims.
    public init(accountID: String? = nil, planType: String? = nil, email: String? = nil) {
        self.accountID = accountID
        self.planType = planType
        self.email = email
    }
}

/// Extracts unverified display and routing claims, preferring the ID token and never authorizing from them.
public func extractUnverifiedClaims(idToken: String? = nil, accessToken: String? = nil) -> UnverifiedClaims {
    let payload = decodeJWTPayload(idToken) ?? decodeJWTPayload(accessToken) ?? [:]
    let authentication = payload["https://api.openai.com/auth"] as? [String: Any] ?? [:]
    return UnverifiedClaims(
        accountID: authentication["chatgpt_account_id"] as? String,
        planType: authentication["chatgpt_plan_type"] as? String,
        email: payload["email"] as? String
    )
}

private func decodeJWTPayload(_ token: String?) -> [String: Any]? {
    guard let token else { return nil }
    let segments = token.split(separator: ".", omittingEmptySubsequences: false)
    guard segments.count > 1,
          let data = base64URLDecode(String(segments[1])),
          let object = try? JSONSerialization.jsonObject(with: data),
          let payload = object as? [String: Any]
    else { return nil }
    return payload
}

File: kotlin/chatgpt-oauth-core/src/main/kotlin/dev/chatgptoauth/Jwt.kt
package dev.chatgptoauth

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

private const val CLAIM_NAMESPACE = "https://api.openai.com/auth"
private val lenientJson = Json { ignoreUnknownKeys = true }

/**
 * Holds JWT metadata that is unverified and never suitable for authorization.
 *
 * @property accountId Optional unverified request-routing identifier.
 * @property planType Optional unverified display plan.
 * @property email Optional unverified display email.
 */
public data class UnverifiedClaims(
    public val accountId: String? = null,
    public val planType: String? = null,
    public val email: String? = null,
)

private fun decodePayload(token: String): JsonObject? {
    val payload = token.split('.').getOrNull(1) ?: return null
    return runCatching {
        lenientJson.parseToJsonElement(decodeBase64Url(payload).toString(Charsets.UTF_8)).jsonObject
    }.getOrNull()
}

/** Extracts unverified display/routing claims; they never authorize, choose a subject, or determine row ownership. */
public fun extractUnverifiedClaims(idToken: String? = null, accessToken: String? = null): UnverifiedClaims {
    val payload = decodePayload(idToken.orEmpty()) ?: decodePayload(accessToken.orEmpty()) ?: JsonObject(emptyMap())
    val auth = runCatching { payload[CLAIM_NAMESPACE]?.jsonObject }.getOrNull()
    return UnverifiedClaims(
        accountId = runCatching { auth?.get("chatgpt_account_id")?.jsonPrimitive?.takeIf { it.isString }?.content }.getOrNull(),
        planType = runCatching { auth?.get("chatgpt_plan_type")?.jsonPrimitive?.takeIf { it.isString }?.content }.getOrNull(),
        email = runCatching { payload["email"]?.jsonPrimitive?.takeIf { it.isString }?.content }.getOrNull(),
    )
}
